tools/fifo.c

```diff
@@ -94,3 +98,83 @@ bool fifo_read(struct fifo *fifo, uint32_t *value)
 
 	return true;
 }
+
+// Start pfifo methods
+
+struct pfifo *pfifo_create(uint16_t capacity)
```

Comment:
Hmm, perhaps we could make the whole thing type agnostic instead of duplicating?

```c
struct fifo *fifo_create(uint16_t capacity, size_t element_size)
```

The `read`, `write` and `peek_back` would use `memcpy()` instead of assignment, and the `values` member would become a `uint8_t data` arbitrary byte array.

It's up to you though!